examples/machine_learning/src/lib.rs

```diff
@@ -138,63 +138,111 @@ fn evaluate_prediction(hits: &mut u32, animal: &Animal, prediction: &[f64]) -> (
-            generate_animal_data(training_set_size, test_set_size);
-
-        oak::print("Training model\n");
-        let mut model = NaiveBayes::<naive_bayes::Gaussian>::new();
-        model
-            .train(&training_matrix, &target_matrix)
-            .expect("failed to train model of dogs");
-
-        oak::print("Predicting\n");
-        let predictions = model
-            .predict(&test_matrix)
-            .expect("failed to predict dogs!?");
-
-        // Score how well we did.
-        let mut hits = 0;
-        let unprinted_total = test_set_size.saturating_sub(10) as usize;
-        for (animal, prediction) in test_animals
-            .iter()
-            .zip(predictions.iter_rows())
-            .take(unprinted_total)
-        {
-            evaluate_prediction(&mut hits, animal, prediction);
-        }
-
-        if unprinted_total > 0 {
-            println!("...");
-        }
-
-        for (animal, prediction) in test_animals
-            .iter()
-            .zip(predictions.iter_rows())
-            .skip(unprinted_total)
-        {
-            let (actual_type, accurate) = evaluate_prediction(&mut hits, animal, prediction);
-            println!(
-                "Predicted: {:?}; Actual: {:?}; Accurate? {:?}",
-                animal.type_, actual_type, accurate
-            );
-        }
-
-        oak::print(&format!(
-            "Accuracy: {}/{} = {:.1}%",
-            hits,
-            test_set_size,
-            (f64::from(hits)) / (f64::from(test_set_size)) * 100.
-        ));
+    fn invoke(&mut self, method_name: &str, _request: &mut oak::Reader, _response: &mut oak::Writer) {
+        oak::print(method_name);
+        match method_name {
+	    "/oak.examples.machine_learning.MachineLearning/Data" => {
+	        
+		oak::print("Data");
+		//(self.training_set_size, self.test_set_size) = (1000, 1000);
+        	// Generate all of our train and test data
+		let (training_matrix, target_matrix, test_matrix, test_animals) =
+            	     generate_animal_data(self.training_set_size, self.test_set_size);
+		 self.training_matrix = Some(training_matrix);
+		 self.target_matrix = Some(target_matrix);
+		 self.test_matrix = Some(test_matrix);
+		 self.test_animals = Some(test_animals);
+	    }
+	    "/oak.examples.machine_learning.MachineLearning/Learn" => {
+	        oak::print("Training model\n");
+		//self.model = NaiveBayes::<naive_bayes::Gaussian>::new();
+		match self.training_matrix {
+		    Some(ref tr) => match self.target_matrix {
+		        Some(ref ta) => self.model
+			    .train(&tr, &ta)
+            	    	    .expect("failed to train model of dogs"),
+                        None => oak::print("target_matrix not set"),
+		    }
+                    None => oak::print("training_matrix not set"),
+		}
+	    }
+	    "/oak.examples.machine_learning.MachineLearning/Predict" => {
+                oak::print("Predicting\n");
+                let mut predictions = None;
+		match self.test_matrix {
```

Comment:
(nit) formatting seems off, consider using https://github.com/rust-lang/rustfmt#quick-start (we should also add it to the setup instructions, and add a presubmit check to verify formatting).

Reply:
This is likely because for a while I was using emacs without rust-mode. Will fix.

Agree about checks!

Reply:
Done